File: 206-reverse-linked-list/reverse-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if(!head || !head->next) return head;

        return reverseLL(head, nullptr);
    }

    ListNode* reverseLL(ListNode* head, ListNode* prev){
        if(!head->next){
            head->next = prev;
            return head;
        }

        ListNode* newHead = reverseLL(head->next, head);
        
        head->next = prev;

        return newHead;
    }
};

File: 206-reverse-linked-list/reverse-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function(head) {
    if(!head || !head.next) return head;

    let prev = null, next = head.next;

    while(next){
        head.next = prev;
        prev = head;
        head = next;
        next = next.next;
    }

    head.next = prev;

    return head;
};